File: FSUIKit/Classes/FSLeftItem.m
//
//  FSLeftItem.m
//  ShareEconomy
//
//  Created by fudon on 2016/9/6.
//  Copyright © 2016年 FudonFuchina. All rights reserved.
//

#import "FSLeftItem.h"

@implementation FSBackItemView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
    }
    return self;
}

- (void)drawRect:(CGRect)rect
{
    [super drawRect:rect];
    
    CGFloat red = 31 / 255.0;
    CGFloat green = 143 / 255.0;
    CGFloat blue = 228.0 / 255.0;
    CGFloat alpha = 1.0;
    if ([_color isKindOfClass:UIColor.class]) {
        [_color getRed:&red green:&green blue:&blue alpha:&alpha];
    }
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetLineCap(context, kCGLineCapRound);
    CGContextSetLineWidth(context, 2);  //线宽
    CGContextSetAllowsAntialiasing(context, true);
    CGContextSetRGBStrokeColor(context, red, green, blue, alpha);  //线的颜色
    
    CGContextBeginPath(context);
    CGContextMoveToPoint(context, self.bounds.size.width * .9, self.bounds.size.width * .1);  //起点坐标
    CGContextAddLineToPoint(context, 1, self.frame.size.height / 2);
    CGContextAddLineToPoint(context, self.frame.size.width * .9, self.frame.size.height * .9);   //终点坐标
    
    CGContextStrokePath(context);
}

@end

@interface FSLeftItem ()

@property (nonatomic,strong) FSBackItemView *backImage;

@end

@implementation FSLeftItem

- (void)setColor:(UIColor *)color
{
    if (_color != color) {
        _color = color;
        
        _backImage.color = color;
        [_backImage setNeedsDisplay];
        _textLabel.textColor = color;
    }
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self leftDesignViews];
    }
    return self;
}

- (void)leftDesignViews
{
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapActionItem)];
    [self addGestureRecognizer:tap];
    
    _backImage = [[FSBackItemView alloc] initWithFrame:CGRectMake(-5 + 5 * (_mode == FSItemTitleModeNOChar), 12, 11, 20)];
    [self addSubview:_backImage];
    
    if (_mode == FSItemTitleModeDefault) {
        _textLabel = [[UILabel alloc] initWithFrame:CGRectMake(_backImage.right + 3, 7, self.bounds.size.width - 14, 30)];
        _textLabel.font = [UIFont systemFontOfSize:15];
        //    _textLabel.textColor = RGBCOLOR(31, 143, 228, 1);
        _textLabel.adjustsFontSizeToFitWidth = YES;
        _textLabel.textColor = [UIColor colorWithRed:0 green:122/255.0 blue:1 alpha:1.0f];
        [self addSubview:_textLabel];
    }
}

- (void)setMode:(FSItemTitleMode)mode
{
    if (_mode != mode) {
        _mode = mode;
        
        _backImage.frame = CGRectMake(-5 + 5 * (_mode == FSItemTitleModeNOChar), 12, 11, 20);
        if (mode == FSItemTitleModeDefault) {
            _textLabel.frame = CGRectMake(_backImage.right + 3, 7, self.bounds.size.width - 14, 30);
        }else{
            if (_textLabel) {
                [_textLabel removeFromSuperview];
                _textLabel = nil;
            }
        }
    }
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    self.alpha = .28;
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint currentPoint = [touch locationInView:self];
    if (!CGRectContainsPoint(self.frame, currentPoint)) {
        [UIView animateWithDuration:.3 animations:^{
            self.alpha = 1;
        }];
    }
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [UIView animateWithDuration:.3 animations:^{
        self.alpha = 1;
    }];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [UIView animateWithDuration:.3 animations:^{
        self.alpha = 1;
    }];
}

- (void)tapActionItem
{
    if (self.tapBlock) {
        self.tapBlock();
    }
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSLeftItem.swift
// FSLeftItemS.swift
// Translated from FSLeftItem.h/m

import UIKit

public enum FSItemTitleMode: Int {
    case `default` = 0
    case noChar = 1  // 没有字符
}

@objcMembers
public class FSBackItemView: UIView {
    
    public var mode: FSItemTitleMode = .default
    public var color: UIColor? {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        backgroundColor = .clear
    }
    
    public override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        var red: CGFloat = 31 / 255.0
        var green: CGFloat = 143 / 255.0
        var blue: CGFloat = 228.0 / 255.0
        var alpha: CGFloat = 1.0
        
        if let c = color {
            c.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        }
        
        guard let context = UIGraphicsGetCurrentContext() else { return }
        context.setLineCap(.round)
        context.setLineWidth(2.5)
        context.setAllowsAntialiasing(true)
        context.setStrokeColor(red: red, green: green, blue: blue, alpha: alpha)
        
        context.beginPath()
        context.move(to: CGPoint(x: bounds.width * 0.9, y: bounds.width * 0.1))
        context.addLine(to: CGPoint(x: 1, y: frame.height / 2))
        context.addLine(to: CGPoint(x: frame.width * 0.9, y: frame.height * 0.9))
        
        context.strokePath()
    }
}

@objcMembers
public class FSLeftItemS: UIView {
    
    private var backImage: FSBackItemView!
    
    public var mode: FSItemTitleMode = .default {
        didSet {
            guard mode != oldValue else { return }
            backImage.frame = CGRect(x: -5 + 5 * (mode == .noChar ? 1 : 0), y: 12, width: 11, height: 20)
            if mode == .default {
                textLabel?.frame = CGRect(x: backImage.frame.origin.x + backImage.frame.width + 3, y: 7, width: bounds.width - 14, height: 30)
            } else {
                textLabel?.removeFromSuperview()
                textLabel = nil
            }
        }
    }
    
    public var color: UIColor? {
        didSet {
            backImage.color = color
            backImage.setNeedsDisplay()
            textLabel?.textColor = color
        }
    }
    
    public var textLabel: UILabel?
    public var tapBlock: (() -> Void)?
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        leftDesignViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        leftDesignViews()
    }
    
    private func leftDesignViews() {
        let tap = UITapGestureRecognizer(target: self, action: #selector(tapActionItem))
        addGestureRecognizer(tap)
        
        backImage = FSBackItemView(frame: CGRect(x: -6.5 + 5 * (mode == .noChar ? 1 : 0), y: 12, width: 11, height: 20))
        addSubview(backImage)
        
        if mode == .default {
            textLabel = UILabel(frame: CGRect(x: backImage.frame.origin.x + backImage.frame.width + 10, y: 7, width: bounds.width, height: 30))
            textLabel?.font = UIFont.systemFont(ofSize: 17, weight: .regular)
            textLabel?.textColor = UIColor(red: 0, green: 122/255.0, blue: 1, alpha: 1.0)
            addSubview(textLabel!)
        }
    }
    
    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        alpha = 0.28
    }
    
    public override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        let currentPoint = touch.location(in: self)
        if !frame.contains(currentPoint) {
            UIView.animate(withDuration: 0.3) {
                self.alpha = 1
            }
        }
    }
    
    public override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        UIView.animate(withDuration: 0.3) {
            self.alpha = 1
        }
    }
    
    public override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        UIView.animate(withDuration: 0.3) {
            self.alpha = 1
        }
    }
    
    @objc private func tapActionItem() {
        tapBlock?()
    }
}
